Core/include/Acts/Propagator/ConstrainedStep.hpp

```diff
@@ -30,137 +31,104 @@ struct ConstrainedStep {
+  void setValue(Scalar value) {
     /// set the accuracy value
-    values[accuracy] = value;
-    // set/update the direction
-    direction = value > 0. ? NavigationDirection::Forward
-                           : NavigationDirection::Backward;
-    return (*this);
+    m_values[accuracy] = value;
   }
 
-  /// Cast operator to double, returning the min/max value
-  /// depending on the direction
-  operator Scalar() const {
-    if (direction == NavigationDirection::Forward) {
-      return (*std::min_element(values.begin(), values.end()));
-    }
-    return (*std::max_element(values.begin(), values.end()));
-  }
+  /// returns the min step size
+  Scalar value() const { return value(currentType()); }
 
   /// Access to a specific value
   ///
   /// @param type is the resquested parameter type
-  Scalar value(Type type) const { return values[type]; }
+  Scalar value(Type type) const { return m_values[type]; }
 
-  /// Return the maximum step constraint
-  /// @return The max step constraint
-  Scalar max() const {
-    return (*std::max_element(values.begin(), values.end()));
+  /// Access to currently leading min type
+  ///
+  Type currentType() const {
+    return Type(std::min_element(
+                    m_values.begin(), m_values.end(),
+                    [](auto a, auto b) { return std::abs(a) < std::abs(b); }) -
+                m_values.begin());
   }
 
-  /// Return the minimum step constraint
-  /// @return The min step constraint
-  Scalar min() const {
-    return (*std::min_element(values.begin(), values.end()));
-  }
+  /// release a certain constraint value
+  ///
+  /// @param type is the constraint type to be released
+  void release(Type type) { m_values[type] = not_set; }
 
-  /// Access to currently leading min type
+  /// Update the step size of a certain type
   ///
-  Type currentType() const {
-    if (direction == NavigationDirection::Forward) {
-      return Type(std::min_element(values.begin(), values.end()) -
-                  values.begin());
+  /// Only navigation and target abortion step size
+  /// updates may change the sign due to overstepping
+  ///
+  /// @param value is the new value to be updated
+  /// @param type is the constraint type
+  /// @param releaseStep Allow step size to increase again
+  void update(Scalar value, Type type, bool releaseStep = false) {
+    if (releaseStep) {
+      release(type);
+    }
+    // check the current value and set it if appropriate
+    // this will also allow signed values due to overstepping
+    if (std::abs(m_values[type]) > std::abs(value)) {
+      m_values[type] = value;
     }
-    return Type(std::max_element(values.begin(), values.end()) -
-                values.begin());
+  }
+
+  void scale(Scalar factor) {
+    assert(factor > 0);
+    m_values[accuracy] = value() * factor;
   }
 
   /// return the split value as string for debugging
-  std::string toString() const;
+  std::string toString() const {
```

Comment:
Just a minor question: I wonder if there is any reason not to implement this as `operator<<`

Reply:
I don't think so - I can change this right away!